src/agora/radio_calibrate.cc

```diff
@@ -602,55 +605,98 @@ bool RadioConfig::InitialCalib(bool sample_adjust) {
     bool good_csi = true;
     long long tx_time(0);
     long long rx_time(0);
+    // Transmit from Beamforming Antennas to Ref Antenna (Down)
     for (size_t i = 0; i < r; i++) {
       if (good_csi == false) {
         break;
       }
+      if (i == ref) continue;
+
+      // Send a separate pilot from each antenna
+      for (size_t ch = 0; ch < cfg_->NumChannels(); ch++) {
+        int tx_flags = SOAPY_SDR_WAIT_TRIGGER | SOAPY_SDR_END_BURST;
+        int ret = ba_stn_[i]->writeStream(
+            this->tx_streams_[i], ch > 0 ? txbuff1.data() : txbuff0.data(),
+            read_len, tx_flags, tx_time, 1000000);
+        if (ret < (int)read_len) {
+          std::cout << "bad write\n";
+        }
+
+        int rx_flags = SOAPY_SDR_WAIT_TRIGGER | SOAPY_SDR_END_BURST;
+        ret = ba_stn_[ref]->activateStream(this->rx_streams_[ref], rx_flags,
+                                           rx_time, read_len);
+
+        Go();
+
+        int flags = 0;
+        std::vector<void*> rxbuff0(2);
+        rxbuff0[0] = buff[cfg_->NumChannels() * i + ch].data();
+        if (cfg_->NumChannels() == 2) {
+          rxbuff0[1] = dummybuff.data();
+        }
+        ret = ba_stn_[ref]->readStream(this->rx_streams_[ref], rxbuff0.data(),
+                                       read_len, flags, rx_time, 1000000);
+        if (ret < (int)read_len) {
+          good_csi = false;
+          std::cout << "bad read (" << ret << ") at node " << ref
+                    << " from node " << i << std::endl;
+        }
+      }
+    }
+    // Transmit from Ref Antenna to Beamforming Antennas (Up)
+    {
       int tx_flags = SOAPY_SDR_WAIT_TRIGGER | SOAPY_SDR_END_BURST;
-      int ret = ba_stn_[i]->writeStream(this->tx_streams_[i], txbuff0.data(),
-                                        cfg_->PilotCi16().size(), tx_flags,
-                                        tx_time, 1000000);
+      int ret =
+          ba_stn_[ref]->writeStream(this->tx_streams_[ref], txbuff0.data(),
+                                    read_len, tx_flags, tx_time, 1000000);
       if (ret < (int)read_len) {
         std::cout << "bad write\n";
       }
-      for (size_t j = 0; j < r; j++) {
-        if (j == i) {
-          continue;
-        }
+
+      for (size_t i = 0; i < r; i++) {
+        if (i == ref) continue;
```

Comment:
For small uses like this, it might be better to just put the three lines in a    if ( i != ref) block